rclc_parameter/src/rclc_parameter/parameter_server.c

```diff
@@ -583,19 +1153,140 @@ rclc_add_parameter(
   }
 
   parameter_server->parameter_list.data[index].value.type = type;
-  parameter_server->parameter_list.size++;
+  ++parameter_server->parameter_list.size;
+
+  // Add to parameter descriptors
+  if (!parameter_server->low_mem_mode && !rclc_parameter_set_string(
+      &parameter_server->parameter_descriptors.data[index].name,
+      parameter_name))
+  {
+    return RCL_RET_ERROR;
+  }
+
+  parameter_server->parameter_descriptors.data[index].type = type;
+  ++parameter_server->parameter_descriptors.size;
 
   if (parameter_server->notify_changed_over_dds) {
-    rclc_parameter_prepare_parameter_event(
+    rclc_parameter_prepare_new_event(
       &parameter_server->event_list,
+      &parameter_server->parameter_list.data[index]);
+    return rclc_parameter_service_publish_event(parameter_server);
+  }
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rclc_add_parameter_undeclared(
+  rclc_parameter_server_t * parameter_server,
+  Parameter * parameter)
+{
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    parameter_server, "parameter_server is a null pointer", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    parameter, "parameter is a null pointer", return RCL_RET_INVALID_ARGUMENT);
+
+  size_t index = parameter_server->parameter_list.size;
+
+  if (index >= parameter_server->parameter_list.capacity ||
+    rclc_parameter_search(&parameter_server->parameter_list, parameter->name.data) != NULL)
+  {
+    return RCL_RET_ERROR;
+  }
+
+  if (RCL_RET_OK != rclc_parameter_copy(
       &parameter_server->parameter_list.data[index],
-      true);
+      parameter))
+  {
+    return RCL_RET_ERROR;
+  }
+
+  ++parameter_server->parameter_list.size;
+
+  // Add to parameter descriptors
+  if (!parameter_server->low_mem_mode && !rclc_parameter_set_string(
+      &parameter_server->parameter_descriptors.data[index].name,
+      parameter->name.data))
+  {
+    return RCL_RET_ERROR;
+  }
+
+  parameter_server->parameter_descriptors.data[index].type =
+    parameter->value.type;
+  ++parameter_server->parameter_descriptors.size;
+
+  if (parameter_server->notify_changed_over_dds) {
+    rclc_parameter_prepare_new_event(
+      &parameter_server->event_list,
+      &parameter_server->parameter_list.data[index]);
     return rclc_parameter_service_publish_event(parameter_server);
   }
 
   return RCL_RET_OK;
 }
 
+rcl_ret_t
+rclc_delete_parameter(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name)
+{
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    parameter_server, "parameter_server is a null pointer", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    parameter_name, "parameter_name is a null pointer", return RCL_RET_INVALID_ARGUMENT);
+
+  if (parameter_server->on_callback) {
```

Comment:
> As you only rely on this single variable, it would be very simple to protect access with a mutex?

The parameter server is not thread safe, nothing stops a `rclc_parameter_get_double` call while the parameter is being deleted by an external client or a simultaneous `rclc_delete_parameter` call.

This is stated on the public API documentation, as all the methods have:
https://github.com/ros2/rclc/blob/eb4b3ca091928a750add48eae0ba0c98e7c06e43/rclc_parameter/include/rclc_parameter/rclc_parameter.h#L167

> As far as I understand, xrcedds suports multi-threading. So does the rclc-executor.

Yes, but the multi-threading protection happens at the `rmw` and `xrcedds` level. We don't have access to the mutex implementation on this layer.